python/lsst/pipe/base/graph/_implDetails.py

```diff
@@ -181,3 +208,77 @@ def __contains__(self, key: _T) -> bool:
             Boolean of the presence of the supplied key
         """
         return key in self._producers or key in self._consumers
+
+
+def _pruner(datasetRefDict: _DatasetTracker[DatasetRef, QuantumNode], refs: Iterable[DatasetRef], *,
+            prunes: Optional[Set[QuantumNode]] = None):
+    if prunes is None:
+        prunes = set()
+    for ref in refs:
+        # make a copy here, because this structure will be modified in
+        # recursion, hitting a node more than once wont be much of an
+        # issue, as we skip anything that has been processed
+        nodes = set(datasetRefDict.getConsumers(ref))
+        for node in nodes:
+            # This node will never be associated with this ref
+            datasetRefDict.removeConsumer(ref, node)
+            if node in prunes:
+                continue
+            # find the connection corresponding to the input ref
+            connectionRefs = node.quantum.inputs.get(ref.datasetType)
+            # Ref must be an initInput ref and we want to ignore those
+            if connectionRefs is None:
```

Comment:
How about:
```
if connectionRefs is None:
    raise RuntimeError(f"Cannot prune on non-Input dataset type {ref.datasetType.name}")
    
```
to be a little more explicit and helpful to users?  (I know we talked about doing nothing here, but it seems just as easy to tell users they screwed up, which is better.)